qemu/tests/vioinput_hotplug.py

```diff
@@ -0,0 +1,77 @@
+import logging
+import time
+
+from virttest import error_context
+from virttest.qemu_devices import qdevices
+from provider import input_tests
+from avocado.core import exceptions
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Test hotplug/unplug of virtio input device
+    1) Boot up w/ one virtio input device
+    2) Unplug one virtio input device
+    3) Hotplug one virtio input device
+    4) Run basic keyboard/mouse test
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def get_device_id(vm):
+        device_list = []
+        for device in vm.devices:
+            if isinstance(device, qdevices.QDevice):
+                if device.get_param("driver") == plug_device:
+                    device_list.append(device)
+        return device_list
+
+    def hotplug_input_dev(vm, dev):
+        error_context.context("Hotplug %s" % dev, logging.info)
+        out, ver_out = vm.devices.simple_hotplug(dev, vm.monitor)
+        if not ver_out:
+            msg = "no % device in qtree after hotplug" % dev
+            raise exceptions.TestFail(msg)
+        logging.info("%s is hotpluged successfully" % dev)
+
+    def unplug_input_dev(vm, dev):
+        error_context.context("Hot-unplug %s" % dev, logging.info)
+        out, ver_out = vm.devices.simple_unplug(dev, vm.monitor)
+        if not ver_out:
+            msg = "Still get %s in qtree after unplug" % dev
+            raise exceptions.TestFail(msg)
+        time.sleep(15)
+        logging.info("%s is unpluged successfully" % dev)
+
+    def run_subtest(sub_test):
+        """
+        Run subtest(e.g. rng_bat,reboot,shutdown) when it's not None
+        :param sub_test: subtest name
+        """
+        error_context.context("Run %s subtest" % sub_test, logging.info)
+        wait_time = float(params.get("wait_time", 0.2))
+        if sub_test == "keyboard_test":
+            input_tests.keyboard_test(test, params, vm, wait_time)
+        elif sub_test == "mouse_test":
+            input_tests.mouse_test(test, params, vm, wait_time, count=1)
+
+    login_timeout = int(params.get("login_timeout", 360))
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=login_timeout)
+    sub_test = params.get("sub_test")
+    plug_device = params["plug_device"]
+
+    # Hotplug an input device
+    new_dev = qdevices.QDevice(plug_device, {'id': 'input0'})
+    hotplug_input_dev(vm, new_dev)
+    time.sleep(5)
+    run_subtest(sub_test)
+    session = vm.reboot(session)
+    # Unplug attached input device
+    unplug_input_dev(vm, get_device_id(vm)[0])
```

Comment:
`vm.devices.simple_unplug`

Reply:
In fact, I also called vm.devices.simple_unplug, just added it in unplug_input_dev().